bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/ValueUtils.java

```diff
@@ -57,19 +75,135 @@ public class ValueUtils {
         try {
-            return valueCreator.createRecordValue(recordTypeName);
+            return getPopulatedRecordValue(valueCreator, recordTypeName, providedFields);
         } catch (BError e) {
             // If record type definition not found, get it from test module.
             String testLookupKey = ValueCreator.getLookupKey(packageId, true);
             if (ValueCreator.containsValueCreator(testLookupKey)) {
-                return ValueCreator.getValueCreator(testLookupKey).createRecordValue(recordTypeName);
+                return getPopulatedRecordValue(ValueCreator.getValueCreator(testLookupKey), recordTypeName,
+                        providedFields);
             }
             throw e;
         }
     }
 
+    private static BMap<BString, Object> getPopulatedRecordValue(ValueCreator valueCreator, String recordTypeName,
+                                                                 Set<String> providedFields) {
+        MapValue<BString, Object> recordValue = valueCreator.createRecordValue(recordTypeName);
+        BRecordType type = (BRecordType) TypeUtils.getImpliedType(recordValue.getType());
+        return populateDefaultValues(recordValue, type, providedFields);
+    }
+
+    public static BMap<BString, Object> populateDefaultValues(BMap<BString, Object> recordValue, BRecordType type,
+                                                              Set<String> providedFields) {
+        Map<String, BFunctionPointer<Object, ?>> defaultValues = type.getDefaultValues();
+        if (defaultValues.isEmpty()) {
+            return recordValue;
+        }
+        defaultValues = getNonProvidedDefaultValues(defaultValues, providedFields);
+        Strand strand = Scheduler.getStrandNoException();
+        if (strand == null) {
+            try {
+                final CountDownLatch latch = new CountDownLatch(defaultValues.size());
+                populateInitialValuesWithNoStrand(recordValue, latch, defaultValues);
+                latch.await();
+            } catch (InterruptedException e) {
+                throw ErrorCreator.createError(
+                        StringUtils.fromString("error occurred when populating default values"), e);
+            }
+        } else {
+            for (Map.Entry<String, BFunctionPointer<Object, ?>> field : defaultValues.entrySet()) {
+                recordValue.populateInitialValue(StringUtils.fromString(field.getKey()),
+                        field.getValue().call(new Object[]{strand}));
+            }
+        }
+        return recordValue;
+    }
+
+    private static Map<String, BFunctionPointer<Object, ?>> getNonProvidedDefaultValues
+            (Map<String, BFunctionPointer<Object, ?>> defaultValues, Set<String> providedFields) {
+        return defaultValues.entrySet().stream()
+                .filter(entry -> !providedFields.contains(entry.getKey()))
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
```

Comment:
Shall we use `for` loop instead of `stream` because `stream`s are generally slow?